### PR TITLE
Update Kotlin compiler to escape package names

### LINKED ISSUE #13182: Error compiling Kotlin code when field names conflict with top-level package names
**What version of protobuf and what language are you using?**
Version: 23.3
Language: Kotlin

**What operating system (Linux, Windows, ...) and version?**
MacOS

**What runtime / compiler are you using (e.g., python version or gcc version)**
n/a

**What did you do?**
1. Create the following file in path `net/example/v1/example.proto`:
```proto
syntax = "proto3";

package net.example.v1;

message Sample {
  string net = 1;
}
```
2. Generate the code with:
```
$ mkdir -p src/main/{java,kotlin}
$ protoc --java_out src/main/java --kotlin_out src/main/kotlin net/example/v1/example.proto
```
3. Compile the code with Maven/Gradle/etc. Here's the pom.xml I used: https://gist.github.com/pkwarren/ea8315f579dc0b2b5ebd362ea46e75d7

**What did you expect to see**
No error compiling generated code.

**What did you see instead?**
```
$ ./mvnw compile
...
[ERROR] Failed to execute goal org.jetbrains.kotlin:kotlin-maven-plugin:1.8.22:compile (compile) on project example: Compilation failure: Compilation failure: 
[ERROR] /example/src/main/kotlin/net/example/v1/SampleKt.kt:[50,7] Unresolved reference: example
[ERROR] /example/src/main/kotlin/net/example/v1/SampleKt.kt:[50,65] Can't inline 'block' here: it may contain non-local returns. Add 'crossinline' modifier to parameter declaration 'block'
[ERROR] /example/src/main/kotlin/net/example/v1/SampleKt.kt:[50,70] No value passed for parameter 'p1'
[ERROR] -> [Help 1]
```

The following change to the generated code appears to resolve the issue - I'm guessing the Kotlin compiler needs to be updated to use backticks for fully qualifying package names:

```diff
diff --git a/src/main/kotlin/net/example/v1/SampleKt.kt b/src/main/kotlin/net/example/v1/SampleKt.kt
index f1f9dd4..f703d05 100644
--- a/src/main/kotlin/net/example/v1/SampleKt.kt
+++ b/src/main/kotlin/net/example/v1/SampleKt.kt
@@ -47,5 +47,5 @@ public object SampleKt {
 }
 @kotlin.jvm.JvmSynthetic
 public inline fun net.example.v1.Example.Sample.copy(block: net.example.v1.SampleKt.Dsl.() -> kotlin.Unit): net.example.v1.Example.Sample =
-  net.example.v1.SampleKt.Dsl._create(this.toBuilder()).apply { block() }._build()
+  `net.example.v1`.SampleKt.Dsl._create(this.toBuilder()).apply { block() }._build()
 
```

**Anything else we should know about your project / environment**

### PR DESCRIPTION
**Pull request by pkwarren**
The current Kotlin code generator creates code which fails to compile with imports like:

```
syntax = "proto3";
package net.example.v1;
message Sample {
  string net = 1;
}
```

This results in a condition where 'net' is ambiguous in some cases (could refer to the field or the package). To avoid these issues, the code generator should enclose package paths in backticks:

```
net.example.v1.SampleKt => `net.example.v1`.SampleKt
```

Fixes #13182.

### CODE REVIEW

**Comment by acozzette** [comment]
@pkwarren Could you merge the current `main` into your branch (or rebase if you prefer)? I think the test failures are things that were fixed recently on main but those fixes might not take effect until you update your branch.

**Comment by pkwarren** [comment]
> @pkwarren Could you merge the current `main` into your branch (or rebase if you prefer)? I think the test failures are things that were fixed recently on main but those fixes might not take effect until you update your branch.

Updated.

**Review by zhangskz — APPROVED** [review]

**Comment by jhump** [comment]
@zhangsk, @fowles, anything else needed before this PR can be merged?

**Comment by zhangskz** [comment]
Thanks for checking in -- looks like this got stuck when going through the internal submit process due to some failing tests that seem to be unrelated.

Giving it a kick but please feel free to ping again if you don't see this submitted soon. Sorry for the delay!

**Comment by pkwarren** [comment]
@zhangskz, @fowles, @acozzette - Anything remaining I need to do on my end to get this PR merged? Happy to merge latest main/rebase+squash if needed.

**Review by zhangskz — APPROVED** [review]

**Comment by zhangskz** [comment]
Looks like the change for this somehow got garbage collected internally after getting stuck on already failing tests -- we gave it another kick and hopefully it should go through properly this time.

**Comment by jhump** [comment]
@zhangskz, any updates? Can this be merged yet?

**Comment by pkwarren** [comment]
I've gone ahead and rebased against latest main and squashed down to a single commit (in case that was preventing this from being automatically merged).

**Pull request closed without merging**
